### PR TITLE
frontend: Warn when multiple rate limiters configured

### PR DESCRIPTION
**Pull request by ryanslade**
We'll display a warning if more than one rate limiter is configured for a single code host:

Screenshot 2020-05-18 at 15 54 37

**NOTE**: The issues below have now been solved by using the `ValidateConfig` in `frontend/db`

There are a couple of issues which are not blocking, but would be nice to solve later:

One, when saving external service config the warning is not displayed immediately, only when you navigate away from the page or refresh.

Two, we need to load rate limit config in both repo-updater, to actually enforce the limits and now the frontend to display warnings. I moved common code to the `extsvc` package to help with this. However, we can't do the DB call in `extsvc` since we have two db packages, one from frontend and one from repo-updater. Importing either in `extsvc` leads to an import cycle. Also, both `frontend` and `repo-updater` have their own definitions for the `ExternalService` struct so we need to pull out the common fields and pass those to the `extsvc` package.

(As a side note, we actually have a third definition of `ExternalService` here: https://github.com/sourcegraph/sourcegraph/blob/8bea5eb42c72423a163932d6f31d9b5cf9ebb21d/internal/api/api.go#L118)

Ideally we should consolidate the different definitions of external service to a central package but I don't think we should do that as part of this PR.

Closes: https://github.com/sourcegraph/sourcegraph/issues/10026

### CODE REVIEW

**Comment by eseliger** [comment]
> One, when saving external service config the warning is not displayed immediately, only when you navigate away from the page or refresh.

To address that, would it be simple to add it also to the `warning` property in the GQL resolver of `updateExternalService`?

**Comment by ryanslade** [comment]
> > One, when saving external service config the warning is not displayed immediately, only when you navigate away from the page or refresh.
> 
> To address that, would it be simple to add it also to the `warning` property in the GQL resolver of `updateExternalService`?

That was actually what I tried first, see: https://github.com/sourcegraph/sourcegraph/issues/10026#issuecomment-627444625

**Comment by eseliger** [comment]
I see, one answer to (1) would be to use markdown in the warning, it's already rendered as markdown in the webapp. Ex:

```md
# Error one
This is really serious!

# Error two
Another veery bad problem
```

```go
type warning struct {
    warning string
}
func (w *warning) AddWarning(title, body string) {
    w.warning += fmt.Sprintf("# %s\n%s\n", title, body)
}
```

I agree with that it should _also_ be a banner, as this is a bad config error. Would be nice to immediately include it in there as well, though. Especially because it's immediately visible to the user :) 

But that won't block this PR, we can file a separate ticket for it to keep track of it.

Deferring the actual review to core people

**Comment by mrnugget** [comment]
What's the reasoning against using this?

https://github.com/sourcegraph/sourcegraph/blob/1800cf25e4902d817bb47ee1a3774d68ffa9b6fd/cmd/frontend/db/external_services.go#L73-L75

It's used when creating

https://github.com/sourcegraph/sourcegraph/blob/1800cf25e4902d817bb47ee1a3774d68ffa9b6fd/cmd/frontend/db/external_services.go#L215-L219

and when updating an external service in the database:

https://github.com/sourcegraph/sourcegraph/blob/1800cf25e4902d817bb47ee1a3774d68ffa9b6fd/cmd/frontend/db/external_services.go#L237-L251

The errors are displayed in the warning automatically. No string formatting etc. needed, since the multierror output is displayed correctly.

File: internal/extsvc/types.go [hunk 1]
```diff
@@ -113,3 +116,116 @@ func WebhookURL(kind string, externalServiceID int64, externalURL string) (strin
...
+// internal/api
+// Until we simplify this, RateLimits below will ask for just the fields it needs
+type Common struct {
+	Config      string
+	Kind        string
+	DisplayName string
+}
+
+// RateLimits returns rate limit config for the supplied external services
+func RateLimits(services []Common) ([]RateLimitConfig, error) {
```
Comment (mrnugget) [inline internal/extsvc/types.go:133]: Does this need to take a slice? If you change it to

```suggestion
func ExtractRateLimitConfig(displayName, kind, config string) (RateLimitConfig, error) {
```

and move the loops to the callers, you can get rid of the whole `Common` thing which I don't think we should introduce.

Comment (ryanslade) [reply internal/extsvc/types.go:133]: Good idea, I initially wanted to encapsulate the logic in the `extsvc` package, including the check if the err is a `ErrRateLimitUnsupported`. I think I can still do that by returning a bool indicating if rate limiting is supported for the host.

Comment (ryanslade) [reply internal/extsvc/types.go:133]: Actually no, that gets confusing. I'll just document it.

**Comment by ryanslade** [comment]
> What's the reasoning against using this?

I didn't know about it 😄 

When I asked about how to tackle this I was pointed towards the approach I've taken in this PR: https://github.com/sourcegraph/sourcegraph/issues/10026#issuecomment-627448627

I'll take a look into your suggested approach now

**Comment by ryanslade** [comment]
@mrnugget PTAL

File: internal/extsvc/types.go [hunk 2]
```diff
@@ -113,3 +116,98 @@ func WebhookURL(kind string, externalServiceID int64, externalURL string) string
...
+// RateLimitConfig represents the internal rate limit configured for an external service
+type RateLimitConfig struct {
+	BaseURL     string
+	DisplayName string
+	Limit       rate.Limit
+	IsDefault   bool
+}
+
+// GetLimitFromConfig gets RateLimitConfig from an already parsed config schema.
+func GetLimitFromConfig(kind string, config interface{}) (rlc RateLimitConfig, err error) {
```
Comment (unknwon) [inline internal/extsvc/types.go:146]: 😄 This file has 0% test coverage, it's a good chance to add some for this function.

Comment (ryanslade) [reply internal/extsvc/types.go:146]: Good point, I'll add some now

**Comment by ryanslade** [comment]
@unknwon Up to 58% now

**Pull request merged**